### ISSUE TITLE
Remote: Iteration Plan for March 2020

### ISSUE THREAD

**Issue by kieferrm** [microsoft/vscode-remote-release#2493]
This plan captures our work on the Remote development extensions in **March**. We follow the same iteration cycle as [VS Code](https://github.com/microsoft/vscode/issues/92242). Although we plan for a whole iteration, we not only ship at the end of an iteration but throughout.

## Plan Items
Below is a summary of the top level plan items.

Legend of annotations:

| Mark                | Description                                       |
| ------------------- | ------------------------------------------------- |
| :runner:            | work in progress                                  |
| :hand:              | blocked task                                      |
| :muscle:            | stretch goal for this iteration                   |
| :red_circle:        | missing issue reference                           |
| :large_blue_circle: | more investigation required to remove uncertainty |
| :black_circle:      | under discussion within the team                  |

### Remote - Core
- [ ] :runner: Can't drag files from a remote vscode window to a local window https://github.com/microsoft/vscode-remote-release/issues/1149 @isidorn

### Remote - SSH
- None.

### Remote - Containers
- [x] Attach to Kubernetes https://github.com/microsoft/vscode-remote-release/issues/2497 @chrmarti
- [x] Explore WSL2 support for Remote - Containers https://github.com/microsoft/vscode-remote-release/issues/2498 @chrmarti
- [x] Investigate removing dependency on Dockerode https://github.com/microsoft/vscode-remote-release/issues/2481 @chrmarti
- [x] Improve proxy support for remote running containers https://github.com/microsoft/vscode-remote-release/issues/2487 @chrmarti
- [x] Respect user defined project name https://github.com/microsoft/vscode-remote-release/issues/222 @chrmarti
- [x] Investigate improving progress shown when creating a dev container https://github.com/microsoft/vscode-remote-release/issues/2499 @chrmarti
- [x] Open repository in a Container, offer to create a `devContainer.json` when it doesn't exist already https://github.com/microsoft/vscode-remote-release/issues/1570 @egamma
- [x] Checkout a GitHub PR or branch into a container  https://github.com/microsoft/vscode-remote-release/issues/2461 @egamma
- [x] New Python 3, 3.6, 3.7, and 3.8 pre-built images and devcontainer.json @chuxel

### Remote - WSL
- None.

-----
### Deferred Items
- Polish Remote Explorer Header https://github.com/microsoft/vscode/issues/86037 @alexr00 @misolori
- Explore speeding up container creation by caching extensions locally @sandy081
- Explore using the git installed on Windows https://github.com/microsoft/vscode-remote-release/issues/1844 @aeschli

**Comment by enjibby** [comment]
I have recently received a notification that #222 is now closed. Will the changes for this make it into the March 2020 work's release?

**Comment by chrmarti** [comment]
@enjibby Yes. (You can also tell by the milestone being set to March.)

**Comment by enjibby** [comment]
You guys are total legends :)

**Comment by rawtaz** [comment]
Here's a few questions about lacking information:

- This iteration plan was for March 2020. Is it not supposed to be over, and now by the 6th of april a new release containing the changes in this iteration plan having been made?

- If yes, where is that release, which version number did it get?

- In general, where can one find a changelog for the release versions of this extension? Looking at https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers there's no such thing, nor is there any changelog or release notes in this repository.

- And finally, I am running VS Code 1.43.2 (updated yesterday), with all extensions updated (yesterday), and when I hit "Check for Extension Updates" it checks and then tells me everything is up to date. If I instead right-click the "Remote - Containers" extension and select "Install Another version…", the most recent version I'm presented with is 0.106.0, which I already have installed. So, why on earth isn't the on https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers listed version 0.110.0 offered to my VS Code?

There's a severe lack of information on these topics.

**Comment by Chuxel** [comment]
@rawtaz Thanks for asking for clarifications. These extension follow the VS Code release cycle and planning process. Dates in the the [VS Code plan](https://github.com/microsoft/vscode/issues/92242) of note:

* March 30th, 2020: Endgame begins
* April 3rd, 2020: Endgame done

Note that, as described [here](https://github.com/microsoft/vscode/wiki/Development-Process#inside-an-iteration), endgame week is finalizing development, not release week.

We are currently in "week 1" of the next development iteration which is planning, bug debt, and release. The iteration plan is published once the plan is complete and this one will be closed.

For versions, there are two "channels" for updates - stable and insiders.  VS Code - Insiders gets extension updates faster than stable.  In terms of version numbers, VS Code and each extension has different version numbers.  The final version number of the extension is not known until the time of release since updates are made to the extensions throughout the iteration to VS Code Insiders only. This allows people to get access to the latest features when they want, or to wait for the stable release. 

Version 0.110 you see is one of these insiders-only releases to allow to allow those that are interested to preview and provide feedback on new features.

An easy way to tell when a stable release has happened is that your version of VS Code itself will update and you'll see release notes appear - including references to the VS Code Remote extensions.

Note that we also ship out-of-band updates to the Remote extensions when critical issues are encountered like the release of 0.51 for Remote - SSH.

**Comment by rawtaz** [comment]
@Chuxel That was a very well written and clear reply. Thanks so much for that great communication! It clears it up quite a bit.

Here are a couple of suggestions I'd like to make which would improve the information to users:

- On the extension details (not just for this one, but in general), please change `Version 
 0.110.0` to something like `Version 
 0.110.0 (insider), 0.106.0 (stable)` - this way people will clearly get a hint as to why they might not get the latest (insider) version in their VS Code, and gather more information about it through the next point:

- Preferably also make the text snippets "insider" and "stable" in the previous suggestion link to a page that contains information about the different channels - this way people who don't know about it will see that there's something to click on, and eventually learn about the different channels. The "insider" one could like to e.g. [this](https://code.visualstudio.com/Docs/supporting/FAQ#_prerelease-versions) or [this](https://code.visualstudio.com/insiders/). The "stable" one I'm not sure.

One thing I didn't see answered is where to find a changelog/release notes with what changes were made in each version. It should be easily noticeable and accessible on e.g. the extension's information page (e.g. the one I linked to before), and preferably also in this repository (considering it's specifically about releases of the Remote extension). can you elaborate?

Finally, wouldn't it be a good idea to just have a setting in VS Code where you can toggle if you want to be on the insider or stable channel (requiring reboot and update of course)?

**Comment by Chuxel** [comment]
> changelog/release notes with what changes were made in each version.

These are in VS Code docs and linked from the main VS Code Release notes. For example, here's the section on VS Code [release notes for 1.43](https://code.visualstudio.com/updates/v1_43#_remote-development) which links to more [detailed release notes here](https://github.com/microsoft/vscode-docs/blob/master/remote-release-notes/v1_43.md).

We do not currently include the extension versions in these release notes since they are quite tied to a VS Code version, but we could do that if that would help.

> Wouldn't it be a good idea to just have a setting in VS Code where you can toggle if you want to be on the insider or stable channel 

@rawtaz The remote extensions frequently depend on updates to VS Code itself which is why they are shipped on the same schedule and features are held out of stable until a VS Code release. Note that VS Code stable and insiders can be installed next to each other on the same machine to make it easy to switch between the two.

**Comment by rawtaz** [comment]
Oh, so.. You pretty much always ship new VS Code versions and new extension versions (for this extension anyway) at the same time? They're not separately released?

Then one will have to practically wait for VS Code updates rather than updates to the individual extensions.

**Comment by Chuxel** [comment]
@rawtaz Yep, the extensions will update when a new version of VS Code ships.

VS Code stable is roughly ever 4 weeks. VS Code Insiders updates daily, so it gets updates more frequently for the remote extensions. That said, VS Code Insiders is what the team itself uses to build the product.
